libpromises/class.c

```diff
@@ -244,11 +242,20 @@ ClassTableIterator *ClassTableIteratorNew(const ClassTable *table, const char *n
 
 Class *ClassTableIteratorNext(ClassTableIterator *iter)
 {
-    void *key_ref = NULL;
-    Class *cls = NULL;
+    MapKeyValue *keyvalue;
 
-    while (RBTreeIteratorNext(iter->iter, &key_ref, (void **)&cls))
+    while ((keyvalue = MapIteratorNext(&iter->iter)) != NULL)
     {
+        Class *cls = keyvalue->value;
+
+        /* Make sure we never store "default" as namespace in the ClassTable,
+         * instead we have always ns==NULL in that case. */
+        CF_ASSERT_FIX(cls->ns == NULL ||
+                      strcmp(cls->ns, "default") != 0,
+                      (cls->ns = NULL),
+                      "Class table contained \"default\" namespace,"
+                      " should never happen!");
```

Comment:
On Wed, Mar 2, 2016 at 9:44 AM, Kristian Amlie <notifications@github.com>
wrote:

> In libpromises/class.c
> <https://github.com/cfengine/core/pull/2506#discussion_r54693749>:
>
> >      {
> > +        Class *cls = keyvalue->value;
> > +
> > +        /* Make sure we never store "default" as namespace in the ClassTable,
> > +         * instead we have always ns==NULL in that case. */
> > +        CF_ASSERT_FIX(cls->ns == NULL ||
> > +                      strcmp(cls->ns, "default") != 0,
> > +                      (cls->ns = NULL),
> > +                      "Class table contained \"default\" namespace,"
> > +                      " should never happen!");
>
> I'm confused about this part. Here you say that you don't store the
> namespace, but you *do* store the namespace specifically in
> ClassTablePut(), by using fullname. Am I missing something?
>
>
"fullname" is the key, in this key-value store, and is indeed always
including the namespace. However this assertion ensures that the *value* of
the key-value store, i.e. struct Class, is in the correct form.

BTW this assertion is rather expensive. I'm contemplating on introducing
something like CF_DEBUG_ASSERT that completely disappears ifdef NDEBUG.